services/load/service.go

```diff
@@ -0,0 +1,666 @@
+		}
+	}
+	return nil
+}
+
+func (s *Service) loadHandler(f string) error {
+	file, err := os.Open(f)
+	defer file.Close()
+	if err != nil {
+		return fmt.Errorf("failed to open file %v: %v", f, err)
+	}
+
+	data, err := ioutil.ReadAll(file)
+	if err != nil {
+		return fmt.Errorf("failed to read file %v: %v", f, err)
+	}
+
+	var o client.TopicHandlerOptions
+	switch ext := path.Ext(f); ext {
+	case ".yaml", ".yml":
+		if err := yaml.Unmarshal(data, &o); err != nil {
+			return errors.Wrapf(err, "failed to unmarshal yaml task vars file %q", f)
+		}
+	case ".json":
+		if err := json.Unmarshal(data, &o); err != nil {
+			return errors.Wrapf(err, "failed to unmarshal json task vars file %q", f)
+		}
+	default:
+		return errors.New("bad file extension. Must be YAML or JSON")
+	}
+
+	l := s.cli.TopicHandlerLink(o.Topic, o.ID)
+	handler, _ := s.cli.TopicHandler(l)
+	if handler.ID == "" {
+		_, err := s.cli.CreateTopicHandler(s.cli.TopicHandlersLink(o.Topic), o)
+		if err != nil {
+			return err
+		}
+	} else {
+		_, err := s.cli.ReplaceTopicHandler(l, o)
+		if err != nil {
+			return err
+		}
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.handlers[path.Join(o.Topic, o.ID)] = true
+	if err := s.items.Set(newTopicHandlerItem(o.Topic, o.ID)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) removeMissing() error {
+
+	if err := s.removeTasks(); err != nil {
+		return err
+	}
+
+	if err := s.removeTemplates(); err != nil {
+		return err
+	}
+
+	if err := s.removeHandlers(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) removeTasks() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	loadedTasks, err := s.loadedTasks()
+	if err != nil {
+		return err
+	}
+
+	for _, id := range diff(s.tasks, loadedTasks) {
+		l := s.cli.TaskLink(id)
+		if err := s.cli.DeleteTask(l); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *Service) removeTemplates() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	loadedTemplates, err := s.loadedTemplates()
+	if err != nil {
+		return err
+	}
+	for _, id := range diff(s.templates, loadedTemplates) {
+		l := s.cli.TemplateLink(id)
+		if err := s.cli.DeleteTemplate(l); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Service) removeHandlers() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	loadedHandlers, err := s.loadedHandlers()
+	if err != nil {
+		return err
+	}
+	for _, id := range diff(s.handlers, loadedHandlers) {
+		pair := strings.Split(id, "/")
+		if len(pair) != 2 {
+			return errors.New("expected id to be topicID/handlerID")
+		}
+		l := s.cli.TopicHandlerLink(pair[0], pair[1])
+		if err := s.cli.DeleteTopicHandler(l); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func diff(m map[string]bool, xs []string) []string {
+	diffs := []string{}
+
+	for _, x := range xs {
+		if m[x] {
+			continue
+		}
+		diffs = append(diffs, x)
+	}
+
+	return diffs
+}
+
+func (s *Service) loadedTasks() ([]string, error) {
+	items, err := s.items.List("tasks")
```

Comment:
We should create constants for the `tasks` , `templates` and `handlers` strings, since they are part of the key for the DOA. A typo would corrupt the data :(.

Reply:
fixed.